# Лабораторная работа №3: практическое применение библиотеки символьных вычислений Python (SymPy).

Выполнила студентка группы 9383 Лихашва Анастасия. Вариант 9.

## Цель работы:

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Задачи лабораторной работы:

1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

## Диофантово уравнение

Вариант 49.

Решить диофантово уравнение: 
### 1204x+ 1029y = 49

In [9]:
import sympy
from sympy.solvers.diophantine import diophantine
from sympy import sympify, symbols

x, y = symbols("x, y", integer=True)
diophantine(sympify("1204*x + 1029*y - 49"))



{(147*t_0 - 329, 385 - 172*t_0)}

Решение: 

$x = 147*t_0 - 329$

$y = 385 - 172*t_0$

## RSA кодирование

Вариант 49.

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 9,m = 33). Шифрованное сообщение имеет вид (3; 31; 20; 10). Подберите закрытую часть ключа и прочитайте исходное слово. Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [14]:
from sympy.crypto.crypto import decipher_rsa, rsa_public_key, rsa_private_key

e = 9
m = 33

key = rsa_private_key(m, e)

code = [3, 31, 20, 10]

alphabet = ['а', 'б', 'в', 'г', 'д', 'е', 'ж', 'з', 'и', 'й', 'к', 'л', 'м', 'н', 'о', 'п',
      'р', 'с', 'т', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'щ', 'ы', 'ь', 'э', 'ю', 'я']

word = []
for i in code:
    word.append(alphabet[decipher_rsa(i, key)-2])

word = ''.join([str(i) for i in word])
print('Исходное слово: ',word)

Исходное слово:  ноги


## Преобразование булевой функции

Вариант 69.

f(x, y, z) = ((y ∨ z)(z xor x)) ∨ ((x xor z)(zy))

In [18]:
func = sympify("((y|z)&(z ^ x))|((x ^ z)&(z&y))", convert_xor=False)
func_DNF = sympy.to_dnf(expr, True)
func_CNF = sympy.to_cnf(expr, True)
print("ДНФ: ", func_DNF)
print("КНФ: ", func_CNF)

ДНФ:  (z & ~x) | (x & y & ~z)
КНФ:  (x | z) & (y | z) & (~x | ~z)


## Выводы

В данной лабораторной работе были практически изучены и применены библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.